### PR TITLE
Improving utilities

### PR DESCRIPTION
**Pull request by szg251**
- TokenName from_string and from_bytes
- Use impl_ops instead of hand-written macros
- Use owned Value as base case to avoid cloning (BREAKING CHANGE for Value utility methods)
- Add some methods to AssocMap
- Implement PartialOrd and Ord for Extended

### CODE REVIEW

File: plutus-ledger-api/src/v1/value.rs [hunk 1]
```diff
@@ -105,6 +318,16 @@ impl TokenName {
     }
 }
 
+impl TokenName {
+    pub fn from_bytes(bytes: Vec<u8>) -> Self {
+        TokenName(LedgerBytes(bytes))
```
Comment (chfanghr) [inline plutus-ledger-api/src/v1/value.rs:323]: Hmmm - `from_string` below implies that a token name stores some utf-8 encoded bytes, should we also validate here? Or perhaps it'd be a great idea to introduce a newtype instead?

File: plutus-ledger-api/src/v1/transaction.rs [hunk 2]
```diff
@@ -211,6 +240,21 @@ impl IsPlutusData for TxInInfo {
     }
 }
 
+impl From<(TransactionInput, TransactionOutput)> for TxInInfo {
+    fn from((reference, output): (TransactionInput, TransactionOutput)) -> TxInInfo {
+        TxInInfo { reference, output }
+    }
+}
+
+impl From<(&TransactionInput, &TransactionOutput)> for TxInInfo {
```
Comment (chfanghr) [inline plutus-ledger-api/src/v1/transaction.rs:249]: [nit]: same comment in the v2 module - do we really need this instance? I think we can return something like &'a TxInInfo and make the clone explicit.

File: plutus-ledger-api/src/v1/interval.rs [hunk 3]
```diff
@@ -247,6 +384,52 @@ where
     PosInf,
 }
 
+impl<T> Ord for Extended<T>
+where
+    T: Ord,
+{
+    fn cmp(&self, other: &Self) -> cmp::Ordering {
+        match self {
```
Comment (chfanghr) [inline plutus-ledger-api/src/v1/interval.rs:392]: Why not: `match (self, other)`? Also, what's the reason to implement both `PartialOrd` and `Ord` manually?

File: plutus-ledger-api/src/v1/interval.rs [hunk 4]
```diff
@@ -82,6 +103,122 @@ where
...
+                    LowerBound {
+                        bound: Extended::Finite(start),
+                        closed: lc,
+                    },
+                to:
+                    UpperBound {
+                        bound: Extended::Finite(end),
+                        closed: uc,
+                    },
+            } => {
```
Comment (chfanghr) [inline plutus-ledger-api/src/v1/interval.rs:133]: Should we ensure that `start < end` always hold true?

File: plutus-ledger-api/src/v1/assoc_map.rs [hunk 5]
```diff
@@ -16,6 +16,60 @@ impl<K, V> AssocMap<K, V> {
...
+
+    /// Inserts a key-value pair into the map.
+    ///
+    /// If the map did not have this key present, None is returned.
+    ///
+    /// If the map did have this key present, the value is updated, and the old value is returned. The key is not updated, though; this matters for types that can be == without being identical. See the module-level documentation for more.
+    pub fn insert(&mut self, key: K, value: V) -> Option<V>
+    where
+        K: PartialEq,
+        V: Clone,
```
Comment (chfanghr) [inline plutus-ledger-api/src/v1/assoc_map.rs:28]: `V` doesn't need to be`Clone` I believe - try `std::mem::swap`.

**Review by chfanghr — CHANGES_REQUESTED** [review]
Great work! Thanks for adding documents to the value utilities btw. I left some comments but mostly minor stuff.

File: plutus-ledger-api/src/v2/transaction.rs [hunk 6]
```diff
@@ -3,6 +3,8 @@ use crate::plutus_data::{parse_constr_with_tag, parse_fixed_len_constr_fields};
 use crate::plutus_data::{
     verify_constr_fields, IsPlutusData, PlutusData, PlutusDataError, PlutusType,
 };
+#[cfg(feature = "chrono")]
+pub use crate::v1::transaction::POSIXTimeConversionError;
```
[the remark this answers is not in the document]
Comment (szg251) [reply plutus-ledger-api/src/v2/transaction.rs:7]: I was reexporting everything that is common between v1 and v2, I think it looks a bit cleaner in the client library, if the versions are not mixed up.

File: plutus-ledger-api/src/v1/value.rs (hunk 1 above)
Comment (szg251) [reply plutus-ledger-api/src/v1/value.rs:323]: I only added a `from_string` and not a `to_string` (which might be a future addition)
`from_string` is safe

File: plutus-ledger-api/src/v1/transaction.rs (hunk 2 above)
Comment (szg251) [reply plutus-ledger-api/src/v1/transaction.rs:249]: It was a nice workflow with tx-bakery, when I used `query_utxo`, I got back a `BTree<TransactionInput, TransactionOutput>`, so I could just use `iter().map(TxInInfo::from)` on it. However, I had to make some changes in the query function, so this is not relevant now.

File: plutus-ledger-api/src/v1/interval.rs (hunk 3 above)
Comment (szg251) [reply plutus-ledger-api/src/v1/interval.rs:392]: good idea

File: plutus-ledger-api/src/v1/interval.rs (hunk 4 above)
Comment (szg251) [reply plutus-ledger-api/src/v1/interval.rs:133]: I added this check

File: plutus-ledger-api/src/v1/value.rs (hunk 1 above)
Comment (szg251) [reply plutus-ledger-api/src/v1/value.rs:323]: Now I also added `try_to_string` with returns a Result

**Comment by szg251** [comment]
@chfanghr I have addressed your comments and added a few tests

**Pull request merged**